homeassistant/components/denonavr/config_flow.py

```diff
@@ -0,0 +1,182 @@
+        if user_input is not None:
+            # Get config option that have defaults
+            self.timeout = user_input[CONF_TIMEOUT]
+            self.show_all_sources = user_input[CONF_SHOW_ALL_SOURCES]
+            self.zone2 = user_input[CONF_ZONE2]
+            self.zone3 = user_input[CONF_ZONE3]
+
+            return await self.async_step_connect()
+
+        return self.async_show_form(step_id="settings", data_schema=SETTINGS_SCHEMA)
+
+    async def async_step_connect(self, user_input=None):
+        """Connect to the receiver."""
+        connect_denonavr = ConnectDenonAVR(
+            self.hass,
+            self.host,
+            self.timeout,
+            self.show_all_sources,
+            self.zone2,
+            self.zone3,
+        )
+        if not await connect_denonavr.async_connect_receiver():
+            return self.async_abort(reason="connection_error")
+        receiver = connect_denonavr.receiver
+
+        unique_id = self.construct_unique_id(
+            receiver.model_name, receiver.serial_number
+        )
+        await self.async_set_unique_id(unique_id)
+        self._abort_if_unique_id_configured()
+
+        return self.async_create_entry(
+            title=receiver.name,
+            data={
+                CONF_HOST: self.host,
+                CONF_TIMEOUT: self.timeout,
+                CONF_SHOW_ALL_SOURCES: self.show_all_sources,
+                CONF_ZONE2: self.zone2,
+                CONF_ZONE3: self.zone3,
+                CONF_RECEIVER_ID: unique_id,
+            },
+        )
+
+    async def async_step_ssdp(self, discovery_info):
+        """Handle a discovered Denon AVR.
+
+        This flow is triggered by the SSDP component. It will check if the
+        host is already configured and delegate to the import step if not.
+        """
+        # Filter out non-Denon AVRs#1
+        if (
+            discovery_info.get(ssdp.ATTR_UPNP_MANUFACTURER)
+            not in SUPPORTED_MANUFACTURERS
+        ):
+            return self.async_abort(reason="not_denonavr_manufacturer")
+
+        # Check if required information is present to set the unique_id
+        if (
+            ssdp.ATTR_UPNP_MODEL_NAME not in discovery_info
+            or ssdp.ATTR_UPNP_SERIAL not in discovery_info
+        ):
+            return self.async_abort(reason="not_denonavr_missing")
+
+        model_name = discovery_info[ssdp.ATTR_UPNP_MODEL_NAME]
+        serial_number = discovery_info[ssdp.ATTR_UPNP_SERIAL]
+        self.host = urlparse(discovery_info[ssdp.ATTR_SSDP_LOCATION]).hostname
+
+        unique_id = self.construct_unique_id(model_name, serial_number)
+        await self.async_set_unique_id(unique_id)
+        self._abort_if_unique_id_configured()
+
+        return await self.async_step_settings()
+
+    @staticmethod
+    def construct_unique_id(model_name, serial_number):
+        """Construct the unique id from the ssdp discovery or user_step."""
+        return f"{model_name}-{serial_number}"
```

Comment:
If I understand correctly, for a receiver that doesn't return `device_info` the `unique_id` will be `Unknown-<IP_ADDRESS>`. It seems to me not very unique.

Reply:
I just commited a PR, improving the handeling of the unique_id and retrieving the serial_number.